File: competitive-programming/leet-code/436/findRightInterval.js
var findRightInterval = function(intervals) {
    const n = intervals.length;
    let indexedIntervals = intervals.map((a, i) => [...a, i]).sort((a, b) => a[0] - b[0]);

    let results = new Array(n).fill(-1);

    for (let i = 0; i < n; i++) {
        let targetEnd = indexedIntervals[i][1];
        let originalIdx = indexedIntervals[i][2];
        
        let l = 0;
        let r = n - 1;
        while (l <= r) {
            let mid = Math.floor((l + r) / 2);
            if (indexedIntervals[mid][0] >= targetEnd) {
                results[originalIdx] = indexedIntervals[mid][2];
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
    }

    return results;
};

File: competitive-programming/leet-code/436/find_right_interval.py
from typing import List

def find_right_interval(intervals: List[List[int]]) -> List[int]:
    n = len(intervals)
    
    indexed_intervals = [[interval[0], interval[1], i] for i, interval in enumerate(intervals)]
    indexed_intervals.sort(key=lambda x: x[0])
    
    results = [-1] * n
    
    for i in range(n):
        target_end = indexed_intervals[i][1]
        original_idx = indexed_intervals[i][2]
        
        left = 0
        right = n - 1
        
        while left <= right:
            mid = (left + right) // 2
            if indexed_intervals[mid][0] >= target_end:
                results[original_idx] = indexed_intervals[mid][2]
                right = mid - 1
            else:
                left = mid + 1
    
    return results